### PR TITLE
Subclass SimilarityEncoder of OneHotEncoder

### PR DESCRIPTION
**Pull request by dsleo**
Replace _BaseEncoder by OneHotEncoder

### CODE REVIEW

File: dirty_cat/similarity_encoder.py [hunk 1]
```diff
@@ -2,9 +2,7 @@
 from scipy import sparse
 
 from sklearn.feature_extraction.text import CountVectorizer, HashingVectorizer
-from sklearn.preprocessing._encoders import _BaseEncoder
-from sklearn.utils import check_array
-
+from sklearn.preprocessing._encoders import OneHotEncoder
```
Comment (GaelVaroquaux) [inline dirty_cat/similarity_encoder.py]: Can you import from sklearn.preprocessing rather than the private module.

Comment (dsleo) [reply dirty_cat/similarity_encoder.py]: sure !

**Comment by GaelVaroquaux** [comment]
Sorry, conflicts to resolve.

**Comment by vinzeebreak** [comment]
There has been some problems with the first merge, we will look into it the sooner we can.

File: dirty_cat/similarity_encoder.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 from scipy import sparse
 from sklearn.feature_extraction.text import CountVectorizer, HashingVectorizer
-from sklearn.preprocessing._encoders import _BaseEncoder
+from sklearn.preprocessing import OneHotEncoder
```
Comment (mcuny) [inline dirty_cat/similarity_encoder.py]: @dsleo can I fix this the merge conflict and commit it to your forks master? Like that we can merge this and close the PR :)

Comment (dsleo) [reply dirty_cat/similarity_encoder.py]: @mcuny weird, I did that change as suggested by Gael yesterday but sure go ahead.

**Comment by dsleo** [comment]
Thinking more about this, and trying not to break LSP, I see two options:  
1. subclass `OneHotEncoder` and raise a `NotImplementedError` on the methods not used in the child class.  
2.  subclass `BaseEnconder` as before and copy the `get_feature_names` from `OneHotEncoder`. After all, it's such small copy/paste.   

@GaelVaroquaux @pcerda: what do you think ?

**Comment by GaelVaroquaux** [comment]
> Thinking more about this, and trying not to break LSP,

I wouldn't worry too much about purity, here.

>  1. subclass OneHotEncoder and raise a NotImplementedError on the methods not used in the child class.

That's not very user friendly.

>  2. subclass BaseEnconder as before and copy the get_feature_names from OneHotEncoder. After all, it's such small copy/paste.

I still prefer subclassing OneHotEncoder, because we inherit
automatically the details of its interface. For instance, if scikit-learn
changes the interface, our code breaks, which is good (because then we
can fix it).

I liked your solution :)

**Comment by dsleo** [comment]
Okay fair enough. Only thing is that the del methods only work for removing parameters not attributes of the class and that's why the checks fail ! 
I'm probably missing a trivial solution but I don't see how to do that easily, unless we create an abstract parent class of OneHotEncoder with only the methods we are interested in here and make SimilarityEncoder subclass of that...

**Comment by GaelVaroquaux** [comment]
Well, can give up on removing those attributes. They are quite harmless, no? 

⁣Sent from my phone. Please forgive typos and briefness.​

**Comment by vinzeebreak** [comment]
I did not add the `NotImplementedError` but if necessary, I can add them in a new commit.

**Comment by GaelVaroquaux** [comment]
I resolved the conflict and will merge.

**Comment by GaelVaroquaux** [comment]
Tests passed. Docs look good. Merging.

**Pull request merged**